msbuild/Xamarin.MacDev.Tasks/Tasks/CollectBundleResources.cs

```diff
@@ -97,11 +98,92 @@ bool ExecuteImpl ()
 
 			bundleResources.AddRange (UnpackedResources);
 
-			BundleResourcesWithLogicalNames = bundleResources.ToArray ();
+			var distinctBundleResources = VerifyLogicalNameUniqueness (Log, bundleResources, "BundleResource");
+
+			BundleResourcesWithLogicalNames = distinctBundleResources.ToArray ();
 
 			return !Log.HasLoggedErrors;
 		}
 
+		[return: NotNullIfNotNull (nameof (items))]
+		public static ITaskItem []? VerifyLogicalNameUniqueness (TaskLoggingHelper Log, IEnumerable<ITaskItem>? items, string itemName)
+		{
+			if (items is null)
+				return null;
+
+			var rv = new List<ITaskItem> ();
+			var groupedBundleResources = items.GroupBy (item => item.GetMetadata ("LogicalName"));
+			var reportedItems = new HashSet<string> (); // Keep track of items we've shown warnings for, to not show multiple warnings.
+
+			foreach (var group in groupedBundleResources) {
+				// No/empty LogicalName is not OK.
+				if (string.IsNullOrEmpty (group.Key)) {
+					foreach (var item in group)
+						Log.LogError (7157, item.ItemSpec, MSBStrings.E7157 /* The {0} item '{0}' does not have a 'LogicalName' metadata. */, itemName, item.ItemSpec);
+					continue;
+				}
+
+				// One item per LogicalName is OK.
+				if (group.Count () == 1) {
+					rv.AddRange (group);
+					continue;
+				}
+
+				// More than one item per LogicalName is not good at all.
+				var notBundledInAssembly = group.Where (item => string.IsNullOrEmpty (item.GetMetadata ("BundledInAssembly")));
+				var bundledInAssembly = group.Where (item => !string.IsNullOrEmpty (item.GetMetadata ("BundledInAssembly")));
+				if (notBundledInAssembly.Count () == 1) {
```

Comment:
You are using Count several times in the IEnumerable, that means that the notBundledInAssembly is being iterated several times. Why not do a ToArray and then use Length? Otherwise in the worst case scenario we are iterating over the collection 3 times.